### DISCUSSION TITLE
Reference coordinate system for Euler Angles MTex

### DISCUSSION THREAD

**Discussion by j-puerstl** [mtex-toolbox/mtex#551] [Ask Anything] [answered]
Hello!

I wanted to find out how the reference coordinate system for calculation of the Euler angles in MTex is chosen, but it's not very clear in the documentation (seeing also that the Oxford system uses two separate coordinate systems for map and Euler angles). Does MTex always give the Euler angles with the respect to the screen coordinate system selected?

On another note, I've also noticed that when using the command 
```matlab
rot = rotation.byEuler(180,0,0)
ebsd_rot = rotate(ebsd,rot,'keepXY')
```
the crystal orientation seems to be rotated by 90°, rather than 180°. Could that be a bug? It has worked well before.

Thanks a lot!

**Comment by zmichels** [comment]
Does the rotation work properly if you do it in degrees?

rot = rotation.byEuler(180*degree,0,0)

**Reply by j-puerstl** [reply]
Perfect! My bad!

**Comment by ralfHielscher** [comment, marked as answer]
Hi,

answering your question is not so simple. The point is that MTEX does not calculate Euler angles out of nothing. 
In most cases Euler angles come from some data load by MTEX. In such an situation the reference frame is of course the reference frame defined by the software the provided the data. 

If you found the documentation at https://mtex-toolbox.github.io/EBSDReferenceFrame.html to short / not clear enough could you please try to be more specific what is not so clear. The question of the two different reference frame of several vendors gets quite some attention in the linked document. 

> Does MTex always give the Euler angles with the respect to the screen coordinate system selected?

Yes.

Ralf.

**Reply by j-puerstl** [reply]
I upload data collected with an Oxford instruments system (ctf/cpr). Following the documentation, I tried different combinations, and the following four give a crystal orientation that overlaps with Channel5:
1) 'convertEuler2SpatialReferenceFrame' - x 'east', z 'intoPlane' - rotationbyEuler(180,0,0), 'keepXY'. Example Euler Angles (337,139.7,231.6)
2) 'Apply rotation only to Euler angles' (180,180,0) - x 'east', z 'intoPlane'. Example Euler Angles (337,139.7,231.6)
3) 'Apply rotation only to spatial coordinates' (0,180,0) - x 'east', z 'outOfPlane' - rotationbyEuler(180,0,0), 'keepXY'. Example Euler Angles (203.2,40.2,51.4)
4) 'convertSpatial2EulerReferenceFrame' - x 'east', z 'outOfPlane' - rotationbyEuler(180,0,0), 'keepXY'. Example Euler Angles (203.2,40.2,51.4)
Now I extract the Euler angles from different grains in MTex, and wanted to further use them to calculate an input orientation for VESTA. Here is where I need to know the reference coordinate system for Euler angle calculation in MTex, because otherwise I cannot extract the Miller Indices needed for VESTA accordingly (upward and projection vector).

What got me a bit confused is that the documentation talks about the screen and Euler angle reference system being different for input data from Oxford instruments. As you can see above, simply aligning these two reference systems ('convertSpatial2EulerReferenceFrame' or 'convertEuler2SpatialReferenceFrame'), and choosing the screen coordinate system so that the map orientation aligns with the Channel5 map did not give me the correct crystal orientation. To align everything, I had to separately then rotate the Euler angles (rotationbyEuler(180,0,0), 'keepXY'), which I understood to correspond to a rotation of the entire Euler angle reference system again. So I thought that this (newly rotated) Euler angle reference system would be what the Euler angles displayed in MTex correspond to, which would have been different from the screen coordinate system. I think a simple statement saying that the Euler angles in MTex are calculated with respect to the map/screen coordinate system early on would have done the trick for me.

In any case, thanks a lot, that answers my question!

**Reply by ralfHielscher** [reply]
Hi,

I have some troubles understanding what you did. Some questions:

1. What does it mean "give a crystal orientation that overlaps with Channel5"? 
2. These 4 points you have  result in 4 different setups. Do you apply all of them or do you think they are equal?

> What got me a bit confused is that the documentation talks about the screen and Euler angle reference system being different for input data from Oxford instruments. 

This is often the case. You can run the import without any correction and check whether it is ok.

> As you can see above, simply aligning these two reference systems ('convertSpatial2EulerReferenceFrame' or 'convertEuler2SpatialReferenceFrame'), and choosing the screen coordinate system so that the map orientation aligns with the Channel5 map did not give me the correct crystal orientation. 

That might happen. Currently `'convertSpatial2EulerReferenceFrame'` and `'convertEuler2SpatialReferenceFrame'` only try to correct this. We hope to improve this in the future. 

>To align everything, I had to separately then rotate the Euler angles (rotationbyEuler(180,0,0), 'keepXY'), which I understood to correspond to a rotation of the entire Euler angle reference system again. So I thought that this (newly rotated) Euler angle reference system would be what the Euler angles displayed in MTex correspond to, which would have been different from the screen coordinate system. I think a simple statement saying that the Euler angles in MTex are calculated with respect to the map/screen coordinate system early on would have done the trick for me.

I think you are right, we should make it more clear in the documentation the Euler angles are interpreted with respect to the map coordinate system.

Ralf.

**Reply by j-puerstl** [reply]
The four different setups described above all result in the same crystal orientations within grains to be displayed in MTex + the grain positions in the map to overlap with the map obtained in Channel5. If I choose a different setup, either the spatial coordinates of the MTex map do not overlap with the map obtained from Channel5, or the grain crystal orientations displayed in MTex (upper image below) do not overlap with the crystal orientation I extracted from Channel 5 (lower image below) - the second point I also double checked with Pole Figures to confirm.

![Crystal Shapes_j](https://user-images.githubusercontent.com/73646809/97583265-a240d200-19ee-11eb-8275-2620e2470c1a.jpg)
![Area3](https://user-images.githubusercontent.com/73646809/97583288-a967e000-19ee-11eb-8f2a-304b3293def9.JPG)

Thanks for your answer, that sounds good!

**Comment by zmichels** [comment]
Hello,

This is a common difficulty when working with data collected with Oxford detectors and software. I have collected data on a few different Oxford systems, and the corrections necessary to get identical plots in MTEX are different for each because each system was differently configured. My solution for this is to take note and save images from Oxford software of both the map appearance/orientation and the pole figure orientation. Then I apply rotations to the data in MTEX to make sure that both the map and the orientation data match how they appeared during acquisition. Once you have the corrections figured out for a given configuration, you can use them regularly on your import. As you pointed out, there are a handful of ways to rotate or import that will provide a similar plot; however, they do not all have the same axis/plotting convention (x-east, z-outOfPage, etc.)… and if you have a preferred coordinate orientation, then your options for corrections may be more limited.

For instance, I prefer x-east and y-north (z-outOfPlane), and on the setup that I currently use, I apply the following corrections during every import:

ebsd = rotate(ebsd,rotation('axis',yvector,'angle',180*degree));
ebsd = rotate(ebsd,rotation('axis',zvector,'angle',180*degree),'keepEuler');

As I mentioned, for different setups, I have needed different corrections.

Best,
Zach

**Comment by kilir** [comment]
One note: Make sure, that what Channel5 gives you is indeed the absolute orientation and the windows all show the same coordinate system. There have been instances where this was not the case. It's been discussed in the past.

Oxford coordinate systems are described here https://github.com/oinanoanalysis/h5oina/blob/master/H5OINAFile.md , unfortunately, the old file formats (ctf, cpr/crc) do not save the entire information.

Cheers,
Rüdiger
